File: C/CRC.c
#include <stdio.h>
#include <string.h>

void xor(char *dividend, const char *divisor, int len) {
    for (int i = 0; i < len; i++) {
        dividend[i] = dividend[i] == divisor[i] ? '0' : '1';
    }
}

void calculate_crc(char *message, const char *generator, char *crc) {
    int message_len = strlen(message);
    int generator_len = strlen(generator);
    char temp[128];
    
    strcpy(temp, message);
    for (int i = 0; i < generator_len - 1; i++) {
        temp[message_len + i] = '0';
    }
    temp[message_len + generator_len - 1] = '\0';
    
    for (int i = 0; i <= message_len; i++) {
        if (temp[i] == '1') {
            xor(temp + i, generator, generator_len);
        }
    }

    strcpy(crc, temp + message_len);
}

int check_crc(const char *message, const char *generator) {
    int message_len = strlen(message);
    int generator_len = strlen(generator);
    char temp[128];
    
    strcpy(temp, message);
    
    for (int i = 0; i <= message_len - generator_len; i++) {
        if (temp[i] == '1') {
            xor(temp + i, generator, generator_len);
        }
    }

    for (int i = message_len - generator_len + 1; i < message_len; i++) {
        if (temp[i] == '1') {
            return 0;
        }
    }
    return 1;
}

int main() {
    char message[128], generator[128], crc[128];
    
    printf("Enter the message in binary: ");
    scanf("%s", message);
    printf("Enter the generator polynomial in binary: ");
    scanf("%s", generator);
    
    calculate_crc(message, generator, crc);
    
    printf("The calculated CRC is: %s\n", crc);
    
    strcat(message, crc);
    printf("The transmitted message with CRC is: %s\n", message);
    
    if (check_crc(message, generator)) {
        printf("No error detected.\n");
    } else {
        printf("Error detected.\n");
    }
    
    return 0;
}

File: Python/CRC.py
def xor(a, b):
    result = []
    for i in range(1, len(b)):
        if a[i] == b[i]:
            result.append('0')
        else:
            result.append('1')
    return ''.join(result)

def mod2div(dividend, divisor):
    pick = len(divisor)
    tmp = dividend[0:pick]

    while pick < len(dividend):
        if tmp[0] == '1':
            tmp = xor(divisor, tmp) + dividend[pick]
        else:
            tmp = xor('0'*pick, tmp) + dividend[pick]
        pick += 1

    if tmp[0] == '1':
        tmp = xor(divisor, tmp)
    else:
        tmp = xor('0'*pick, tmp)

    return tmp

def encode_data(data, key):
    l_key = len(key)
    appended_data = data + '0'*(l_key-1)
    remainder = mod2div(appended_data, key)
    codeword = data + remainder
    return codeword

def decode_data(data, key):
    l_key = len(key)
    remainder = mod2div(data, key)
    return remainder

def main():
    while True:
        print("\n--- CRC Error Detection Protocol ---")
        print("1. Encode Data")
        print("2. Decode Data")
        print("3. Exit")
        choice = input("Enter your choice: ")

        if choice == '1':
            data = input("Enter the data in binary: ")
            key = input("Enter the generator polynomial in binary: ")

            encoded_data = encode_data(data, key)
            print(f"Encoded data with CRC: {encoded_data}")

        elif choice == '2':
            received_data = input("Enter the received data in binary: ")
            key = input("Enter the generator polynomial in binary: ")

            remainder = decode_data(received_data, key)
            print(f"Remainder after division: {remainder}")

            if '1' in remainder:
                print("Error found in received data.")
            else:
                print("No error detected in received data.")

        elif choice == '3':
            print("Exiting program.")
            break

        else:
            print("Invalid choice. Please try again.")

# Call the main function directly
main()
